### PR TITLE
MOEN-34219: Logging & Network Timeout Improvements

### PR DESCRIPTION
**Pull request by abhishek-engg-mobilesdk**
### Jira Ticket
https://moengagetrial.atlassian.net/browse/MOEN-34219

### Description
- Adding option to configure retry count for sonatype and timeout duration for all network calls
- Logging improvements

### CODE REVIEW

File: plugin/src/main/kotlin/com/moengage/internal/MavenPublishManager.kt [hunk 1]
```diff
@@ -47,15 +48,29 @@ internal class MavenPublishManager(private val project: Project) {
         releaseVariant = project.findProperty(RELEASE_VARIANT) as? String ?: DEFAULT_RELEASE_VARIANT
         releasePortal = ArtifactReleasePortal.getMavenCentralPortal(project.findProperty(RELEASE_HOST) as? String)
 
-        val serviceBuilder =
-            ServiceBuilder(releasePortal, project.getUserName(releasePortal), project.getUserPassword(releasePortal))
+        val maxRetryCountOnTimeOut =
```
Comment (umangmoe) [inline plugin/src/main/kotlin/com/moengage/internal/MavenPublishManager.kt:51]: Should there be a limit on the maximum retry? like, say 5 or 10? Beyond that would be DDOS on the server if the server down or something

File: plugin/src/main/kotlin/com/moengage/internal/MavenPublishManager.kt [hunk 2]
```diff
@@ -47,15 +48,29 @@ internal class MavenPublishManager(private val project: Project) {
...
+            (project.findProperty(NEXUS_REPOSITORY_MAX_RETRY_ON_TIMEOUT) as? String)?.toInt() ?: 0
+        val networkTimeoutDuration = (project.findProperty(NETWORK_CALL_TIMEOUT_DURATION) as? String)?.toLong() ?: defaultTimeout
```
Comment (umangmoe) [inline plugin/src/main/kotlin/com/moengage/internal/MavenPublishManager.kt:53]: should we have a maximum timeout value? waiting for infinite time or keeping the connection open for a very long time might not make sense.

File: plugin/src/main/kotlin/com/moengage/internal/repository/NexusRepository.kt [hunk 3]
```diff
@@ -10,6 +10,7 @@ import com.moengage.internal.model.NexusStagingRepositoryData
 import com.moengage.internal.repository.network.NexusService
```
Comment (umangmoe) [inline plugin/src/main/kotlin/com/moengage/internal/repository/NexusRepository.kt:10]: `currentRetryCount` could be named `retryCount` across the file.

**Review by umangmoe — CHANGES_REQUESTED** [review]

File: plugin/src/main/kotlin/com/moengage/internal/repository/network/ServiceBuilder.kt [hunk 4]
```diff
@@ -20,23 +22,29 @@ private const val timeout = 60L
...
-            .connectTimeout(timeout, TimeUnit.SECONDS)
-            .readTimeout(timeout, TimeUnit.SECONDS)
-            .writeTimeout(timeout, TimeUnit.SECONDS)
-            .build()
+        val okHttpClient = OkHttpClient.Builder().apply {
+            connectTimeout(networkTimeoutDuration, TimeUnit.SECONDS)
+            readTimeout(networkTimeoutDuration, TimeUnit.SECONDS)
+            writeTimeout(networkTimeoutDuration, TimeUnit.SECONDS)
+            addInterceptor(AuthorizationInterceptor(artifactReleasePortal, username, password))
+            addInterceptor(httpLogger)
```
[the remark this answers is not in the document]
Comment (abhishek-engg-mobilesdk) [reply plugin/src/main/kotlin/com/moengage/internal/repository/network/ServiceBuilder.kt:39]: Moved to custom interceptor with logging message, so log level will be handled by logger

File: plugin/src/main/kotlin/com/moengage/internal/repository/network/ServiceBuilder.kt [hunk 5]
```diff
@@ -20,23 +22,29 @@ private const val timeout = 60L
 internal class ServiceBuilder(
     private val artifactReleasePortal: ArtifactReleasePortal,
     private val username: String,
-    private val password: String
+    private val password: String,
+    private val networkTimeoutDuration: Long
 ) {
 
+    private var httpLogger: HttpLoggingInterceptor = HttpLoggingInterceptor { message ->
```
[the remark this answers is not in the document]
Comment (abhishek-engg-mobilesdk) [reply plugin/src/main/kotlin/com/moengage/internal/repository/network/ServiceBuilder.kt:29]: Removed external dependency

File: plugin/src/main/kotlin/com/moengage/internal/repository/NexusRepository.kt [hunk 6]
```diff
@@ -30,62 +32,118 @@ internal class NexusRepository(
...
-        if (!response.isSuccessful) {
-            log(LogLevel.ERROR, "$tag getRepositoryData(): ${response.errorBody()}")
-            return null
+    private fun <T> retryIfRequired(currentRetryCount: Int, retryFunction: () -> T?): T? {
+        log(LogLevel.ERROR, "$tag retryIfRequired(): Timeout, currentRetryCount = $currentRetryCount, maxRetryCount = $maxRetryCountOnTimeOut")
+        return if (currentRetryCount < maxRetryCountOnTimeOut) {
+            log(LogLevel.ERROR, "$tag retryIfRequired(): Retrying ...")
+            retryFunction()
+        } else {
+            log(LogLevel.ERROR, "$tag retryIfRequired(): Retry limit reached")
```
[the remark this answers is not in the document]
Comment (abhishek-engg-mobilesdk) [reply plugin/src/main/kotlin/com/moengage/internal/repository/NexusRepository.kt:145]: Changed to warning

File: plugin/src/main/kotlin/com/moengage/internal/MavenPublishManager.kt (hunk 1 above)
Comment (abhishek-engg-mobilesdk) [reply plugin/src/main/kotlin/com/moengage/internal/MavenPublishManager.kt:51]: Added max retry as 5

File: plugin/src/main/kotlin/com/moengage/internal/MavenPublishManager.kt (hunk 2 above)
Comment (abhishek-engg-mobilesdk) [reply plugin/src/main/kotlin/com/moengage/internal/MavenPublishManager.kt:53]: Added max timeout duration as 5 min

File: plugin/src/main/kotlin/com/moengage/internal/repository/NexusRepository.kt (hunk 3 above)
Comment (abhishek-engg-mobilesdk) [reply plugin/src/main/kotlin/com/moengage/internal/repository/NexusRepository.kt:10]: done

**Review by umangmoe — APPROVED** [review]

**Pull request merged**